rotary_utils/repair.py

```diff
@@ -415,54 +404,92 @@ def link_contig_ends(contig_record: SeqIO.SeqRecord, bam_filepath: str, length_o
+                        strip_descriptions=True)
+
+    # Save a copy of the final circlator merge logfile in the main log directory
+    shutil.copy(os.path.join(stitch_dirs.length_outdir, 'merge', 'merge.circularise_details.log'),
+                os.path.join(stitch_dirs.log_dir_base, f'{contig_id}_circlator_final.log'))
+
+
 def iterate_linking_contig_ends(contig_record: SeqIO.SeqRecord, bam_filepath: str, linking_outdir: str,
-                                length_thresholds: list, cli_tool_settings_dict: dict,  verbose_logfile: str,
-                                threads: int = 1):
+                                length_thresholds: list, tool_settings: RepairToolSettings, verbose_logfile: str):
     """
     Iterate link_contig_ends to try to stitch the ends of a circular contig using multiple length thresholds.
 
-    :param contig_record: SeqRecord of the circular contig
+    :param contig_record: SeqRecord of the circular contig.
     :param bam_filepath: path to a BAM file with mapping information of long reads to the contig (it is OK if this file
-                         also contains mappings to other contigs outside the contig of interest)
+                         also contains mappings to other contigs outside the contig of interest).
     :param linking_outdir: output directory for the analysis; will be created by the function, although it can exist
-                           before the function is run
-    :param length_thresholds: list of bp regions around the contig ends to attempt to subset for the assembly
-    :param cli_tool_settings_dict: dictionary containing the following CLI tool settings, as defined in main(), as keys:
-                                   flye_read_mode, flye_read_error, circlator_min_id, circlator_min_length,
-                                   circlator_ref_end, circlator_reassemble_end
-    :param verbose_logfile: path to a logfile where shell script logs will be saved
-    :param threads: parallel processor threads to use for the analysis
-    :return: boolean of whether end linkage was successful (True) or not (False)
+                           before the function is run.
+    :param length_thresholds: list of bp regions around the contig ends to attempt to subset for the assembly.
+    :param tool_settings: RepairToolSettings object containing settings for tools used in the end repair workflow.
+    :param verbose_logfile: path to a logfile where shell script logs will be saved.
+    :return: boolean of whether end linkage was successful (True) or not (False).
     """
 
-    os.makedirs(linking_outdir, exist_ok=True)
-
-    # Where important log files will be copied
-    log_dir_base = os.path.join(linking_outdir, 'logs')
+    stitch_dirs = StitchDirectories(linking_outdir, make_dirs=True)
 
     # Keep trying to link the contig ends until successful or until all length thresholds have been attempted
     assembly_attempts = 0
     linked_ends = False
-
     while linked_ends is False:
```

Comment:
@jmtsuji I would double check this to determine there isn't a corner case where you could get a runaway loop.

Reply:
I've checked the code pretty carefully and think a runaway loop would not occur. At every exit point, 1 is added to `assembly_attempts`.